### DISCUSSION TITLE
How to write my own model?

### DISCUSSION THREAD

**Discussion by gcxxx** [opencobra/cobrapy#1141] [Q&A]
Hi everyone:

I have designed a middle size model of my own(about 100 reactions and 70 metabolites) and I have written all chemical reaction with stoichiometric formula.My question is how to translate the model into the format python can recognize? (I know something about sbml,but I am not familiar with that)

**Comment by bdelepine** [comment]
Hi @gcxxx !

I understand that you want to use COBRApy to handle and/or make simulations with your model, but I didn't get in which format your model is right now: could you provide more details? Maybe the tool(s) you used to reconstruct it?

COBRApy can read models in several pseudo-standard formats [as described in the documentation](https://cobrapy.readthedocs.io/en/latest/io.html). SBML should be your go-to format to store/exchange/publish models, but it is not designed to be manually edited (you can do it, but it's easy to make mistakes). Most of the tools you can use to build a metabolic model have a built-in SBML export. Personally, I like [MetExplore](https://vm-metexplore-prod.toulouse.inra.fr/index.html/) to design new models from scratch and make some basic simulations, but other tools exist.

COBRApy can also be used to build a model from a non-standard format (and eventually export it to SBML). This would require some Python knowledge to extract data from the non-standard format and instantiate the right objects [as described in the documentation](https://cobrapy.readthedocs.io/en/latest/building_model.html). If you go this way, you may be interested in [Reaction.build_reaction_from_string](https://cobrapy.readthedocs.io/en/latest/autoapi/cobra/core/reaction/index.html?#cobra.core.reaction.Reaction.build_reaction_from_string) that reads a string such as `A + B => C` and creates automatically the reaction and the metabolites.

Let me note here that designing from scratch a metabolic model is not so common as most often you would either propagate reaction assignation from genomic and homology data (see [MetaNetX](https://www.metanetx.org/cgi-bin/mnxweb/upload-genome), [ModelSEED](https://modelseed.org/), and [apps from KBase](https://kbase.us/applist/#Metabolic%20Modeling) among others), or build onto a pre-existing published model close to your organism of interest (here again, several model repositories exist, but [BiGG](http://bigg.ucsd.edu/) is maybe the most specialized).

Does that answers your question?

**Reply by gcxxx** [reply]
Hi @bdelepine !
Thank you again for your zealous answer!!!! I appreciate that!!
![photo](https://user-images.githubusercontent.com/86390409/147670010-f72c87fb-3318-481b-aec0-b3faecaf146b.png)
As you can see, I have designed them in excel which is most likely not in the format you understand😂 And you are right, all I want to do
is make FBA simulations on a smaller model so that I can analyze the flux distribution more conveniently. Unfortunately, as far as I know, all CHO cell model published are so large in size (commonly thousands of reactions and metabolites) that it is very difficult to analyze the flux distributions, besides, fva result always give me a very ''unconstrained state'' (many reactions with full flux range).

After encountering these difficulties, I come up with an idea of designing my own model (which contains glycolysis, pentose phosphate
pathway (PPP), tricarboxylic acid cycle (TCA) as well as the pathways of glutaminolysis, amino acid metabolism and synthesis of
proteins, nucleotides, lipids and biomass were included with no compartmentalization, thus covering only the relevant metabolic
pathways in this simplified reaction network ).  The number of constrains I can give to the model is 24,which means there are only 6 freedom (100-70-24=6), I hope in this way fva will give me a more 'constrained state' of my model.

Generally speaking, I want to run FBA simutions but I don't know how to turn the excel file into something cobrapy can run.

**Reply by bdelepine** [reply]
I see. Well, to build a simpler model you have the choice to:

1. use a tool to rebuild your model (following your Excel file), then export it to SBML and play with it in COBRApy.
2. or, parse your Excel file with Python and build the model "on-the-fly" with COBRApy.
3. or, go back to the best CHO model available and get rid of all reactions (or constrain their flux to 0) you do not care about.

For 1/, I would use a tool like MetExplore that has a GUI and some built-in simulation and flux visualization features.

For 2/, if you are not afraid of Python you could do something along:

```python
import pandas as pd
import cobra

df = pd.read_excel("path/to/excel/file")
model = cobra.Model('Model name')

for idx, row in df.iterrows():
    # Parse row
    this_rid = row["Reaction ID"]
    this_reactant = row["Reactant"]
    this_product = row["Products"]
    this_direction = row["Direction"]
    this_equation = f"{this_reactant}{this_direction}{this_product}"
    # Create a new reaction
    reaction = cobra.Reaction(this_rid)
    model.add_reactions([reaction])
    reaction.build_reaction_from_string(this_equation)
print(model)
```

For both 1/ and 2/ you would have to carefully design all the required-for-modelling reactions and constraints: biomass reaction, exchange reactions, media, check the mass balance, etc. This may get complicated. 

That's why you may consider 3/ to "simplify" the best model available with something along:

```python
model = cobra.io.read_sbml_model("iCHOv1.xml") 

whitelist_reaction_ids= ['EX_glc__D_e', 'EX_h_e', 'GLCt2', 'HEX1', ..........]

for r in model.reactions:
    if r.id not in whitelist_reaction_ids:
        r.knock_out()
``` 

The 3/ has the advantage to start from something that is already carrying flux, which should be easier to debug.

You may also be interested in tools such as [Escher](https://escher.readthedocs.io/en/latest/index.html) or [cy3sbml app for Cytoscape desktop](https://apps.cytoscape.org/apps/cy3sbml) to visualize your model and its flux distribution.

**Comment by gcxxx** [comment]
@bdelepine First, thank you again for your answer!
I tried method2 (using python to construct) you suggest, something happened unexpectedly.
![photo](https://user-images.githubusercontent.com/86390409/147739243-39ea3b53-fae4-4d70-ac2d-89d98f0ee22f.png)
It seems like model has formed bunch of repetitive metabolites. I upload the excel file.
[gcxx.xlsx](https://github.com/opencobra/cobrapy/files/7792329/gcxx.xlsx)

Besides,**not all metabolites in my model are balanced**, some of them are either formed or consumed.(ie NAD,FAD,ATP,GTP etc),how to handle them?  They should not be one of the metabolites in the metabolite-reaction matrix, is that right?

**Reply by bdelepine** [reply]
Hi gcxx :) 

You're entering into the "you would have to carefully design all the required-for-modelling reactions and constraints" problem ;)

Careful: all metabolic reactions **must** be balanced for your model to make sense. However, as there is no concept of "metabolic pools" in a classic genome-scale metabolic model + FBA framework (unlike kinetic models), you must design "special" (non-metabolic) reactions to create and make disappear atoms from the system. Otherwise, there could be no simulated fluxes (except balanced cycles).

The biomass reaction is usually used to "make disappear" the strict equivalent of 1g of dry cell weight; this is a modelling proxy for "cell growth". Boundary reactions are used to "create" and "make disappear" metabolites in/out of the system; among those, exchange reactions do so in the extracellular space (so metabolites can be transported and metabolized) and are a modelling proxy for "allowed fluxes in/out the cell's environment". You will have to craft your own: e.g. `CO2(e)  <----> ` to allow the model to "create"/"make disappear" CO2(e) from the system.

I strongly encourage you to read:
* https://www.nature.com/articles/nprot.2009.203 to get a clear idea of how to create a metabolic model for FBA-like simulations
* https://www.nature.com/articles/s41587-020-0446-y/ to get best practises

You may also be interested in the great lectures from Palsson available on Youtube. I think [this one](https://www.youtube.com/watch?v=KY7P0IMar9U&list=PLzm7DNlS3_0hryL9WsVbWV_EkwBJNt9s3&index=2) is the most relevant to explain boundary reactions.

That being said, in your example, metabolites are not really duplicated: as the prints shows, the stoichiometric number was recognized as part of the metabolite identifier. You should use a format `A + n B -----> C`, i.e. without "*".

**Reply by gcxxx** [reply]
Thanks a lot, it seems like there are tens of thousands of knowledge needed to be learnt😂I will take some time.

**Reply by gcxxx** [reply]
Hi @bdelepine 

After reading a dozen of FBA-related articles and several times of trying(your advice is very useful!), I successfully built this model using cobrapy !!! The only question remaining is I don't know how to visualize it, my model is built following the suggestion of  **''parse your Excel file with Python and build the model "on-the-fly" with COBRApy**.  By the way, I also found the topology of my reactions matter , (I actually debugged almost hundreds of times😂, it seems like the balance of all metabolites is not the only important factor) I had to examine extremely carefully to make sure fva result of each reaction is not zero, is this related to the EFMs kind of things? 

Besides,
**However, as there is no concept of "metabolic pools" in a classic genome-scale metabolic model + FBA framework (unlike kinetic models), you must design "special" (non-metabolic)**
I am now wondering the topic of kinetic model you proposed, is there a way to combine fba model with kinetic model? I designed 
a kinetic model using ODE (system of differential equations) before and I don't know if they are the same kind of thing or related to some extent. There are some introductions about dynamic FBA on the documentation page, but it is not enough to understand the whole picture. I want to ask you for your advice because my gut tells me you are definitely a PRO in region of this kind😂

**Reply by bdelepine** [reply]
Hi @gcxxx 

I'm happy to know that you succeed! 😄🎉

About the visualization. [This question is in the FAQ](https://cobrapy.readthedocs.io/en/latest/faq.html#How-do-I-visualize-my-flux-solutions?). I would recommend [Escher](https://github.com/zakandrewking/escher) and/or [cy3sbml app for Cytoscape](https://apps.cytoscape.org/apps/cy3sbml) desktop to visualize your model and its flux distribution. The key is to refrain from wanting to display everything on the same view, and to use instead several views of your network.

About the topology. That's not only a story of enumerating the feasible routes (that would give you the EFM). Remember that your flux distribution is the result of an optimization process, usually to maximize biomass. So, when two routes (or a combination of different routes) are equivalents and give some flux variability it means that they are "equivalent in the eye of the objective"... but (for classic FBA) that's not taking into account enzyme kinetics, gene expression, differential cost in enzyme synthesis (e.g. a short and a long pathway having the same metabolic balance are identical for an FBA but may have a significantly different enzyme burden), cofactors that are not part of the reaction equation (e.g. prosthetic groups, ions, etc.), etc.

About the "kinetic models" I was referring to. Yes, I was thinking about a simple set of ODE with rate laws and enzymatic parameters determined _in vitro_ (Km, Kcat, etc.). Something you would build with COPASI for instance. 

About kinetic models and FBA. There are plenty of attempts at using reaction/enzyme's kinetic data in constraint-based metabolic models, even for genome-scale models. Basically, people use both types of models sequentially (like exemplified with [dynamic FBA](https://cobrapy.readthedocs.io/en/latest/dfba.html) that uses the kinetic of substrate assimilation) to add a time-dimension to the simulation; or they find some clever way to translate the enzyme-related-data into constraints that can be understood by the solver used for an FBA, in the hope to get a more accurate flux distribution. For instance, one may explicitly use enzyme kinetic parameters such as kcat (see [this approach](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6325765/)). In the same spirit, one may "only" take into account an idea of enzyme abundance (~Vmax) and constrain reactions' bounds with expression data (see [ME-model](https://pubmed.ncbi.nlm.nih.gov/24084808/)). Going further, you may be interested in [RBApy](https://github.com/SysBioInra/RBApy) that aims to take many growth/production limiting factors into account, including enzyme kinetic data (if I recall correctly).

**Reply by gcxxx** [reply]
Hi @bdelepine  Thank you very much for your reply!😄 Your advice and generous answer mean a lot to beginner like me😀 I want to discuss with you remaining questions. 

First of all, about the visualization: Since I built the model in cobrapy, is there a way to export my model as sbml file using cobrapy?

Secondly, about the topology: When I was debugging my model, I ran fva to see results after adding each reaction to the model to avoid the situation that one or some reactions carried zero flux. Your point is this movement is necessary, but there are other factors needed to be considered, is that right? ''**So, when two routes (or a combination of different routes) are equivalents and give some flux variability it means that they are equivalent in the eye of the objective**''I can't completely understand the definition of 'equivalent routes' ,would you mind to take an example easy enough for me to understand?😂😂

Finally, about the 'FBA+Kinetic': I read several papers about using enzyme constraints(just like you said'' **dynamic FBA that uses the kinetic of substrate assimilation**'' or ''**find some clever way to translate the enzyme-related-data into constraints**'') on FBA to make an accurate simulation,  nevertheless, by far I havn't started to make my own dynamic FBA model(or FBA plus enzyme-related data).Therefore, which of COPASI and RBAPy is easier for beginner to use in your opinion?(It seems like they both need sbml file which is the first but most urgent problem needs to be solved😂)

-----------------------------------------------------------------------------------------------------

''**Remember that your flux distribution is the result of an optimization process, usually to maximize biomass**'' When I was buliding the model, I kept asking myself ''what's the difference FBA and MFA?'' I found even if I bulit a model without setting the optimize reaction, fva could still return me a result. When I set some constraints , I found that  
when number of metabolites equals number of reactions(unconstrained) , fva result would give a unique series of  numbers instead of series of range.  Can I jump to conclusion that FBA==MFA if enough constraints are set?

**Reply by bdelepine** [reply]
Hi gcxx,

Escher accept COBRApy models exported in the JSON format ([Escher doc'](https://escher.readthedocs.io/en/latest/escher_and_cobrapy.html)). COBRApy can export models in SBML (and JSON) as stated in the [documentation](https://cobrapy.readthedocs.io/en/latest/io.html#SBML)).

About your FVA, I am not certain to understand what you did and why 🤔. Reactions do not have to carry flux all the time: it is normal to have reactions with a flux of 0, even more if the associated genes are expected to be repressed in those conditions. What you probably do not want is your model to be unable to carry flux at all (in any reaction), but it's easy enough to check that looking at your objective value.

About the routes. Sorry if it has confused you. I was commenting on the elementary flux modes, and basically I wanted to agree with you that the topology of a metabolic network constrains it. I also wanted to insist on the importance of the objective to interpret a flux distribution: it is not "only" a matter of topology, it's also a matter of metabolic cost to build the objective's precursors. Say that no ATP at all was necessary to build the precursors of your biomass reaction (or that ATP synthesis would be "free" through a sink reaction `ATP <-->`). Then, you would see no differences in flux between two pathways (two routes) that only differ in ATP usage. The problem would not be about the topology of your two pathways, but rather that in your simulation conditions ATP has no cost "in the eye of the objective". So, like always, be careful what you optimize for.

About COPASI vs. RBApy. Careful: mixing kinetic models and GEM is not easy and require solid knowledge of constraint-based modelling. I would advise to focus on learning to create kinetic models (with COPASI) or GEM (with COBRApy) depending on your biological problematic, but not both.

About MFA. It depends to what kind of MFA you are referring to. See [this review](https://pubmed.ncbi.nlm.nih.gov/25613286/).

**Reply by gcxxx** [reply]
Hi @bdelepine Thank you for your reply!
About the model export: I have exported my built on-the-fly successfully as json file, but I met some unexpected error when I try to export it as sbml file
![error](https://user-images.githubusercontent.com/86390409/150642275-8fabfc75-96e8-4384-98ef-a3b76ec35c6c.png)

About the FBA you say, I think I need to take some time to understand your point😂 

About the COPASI: I find that COPASI can build not only kinetic model but also fba model as well, even hybrid model(fba+kinetic).
